File: packages/capabilities-runtime-ts/src/node/filesystem.ts
/**
 * Filesystem adapter foundation (§15.3). Every target path is
 * root-relative, normalized, and checked against traversal and symlink
 * escape before any I/O happens; read and write are separate, explicitly
 * granted capabilities so a read-only composition can never accidentally
 * write.
 */

import { promises as fs } from 'node:fs'
import * as path from 'node:path'

export class FilesystemPathError extends Error {
  constructor(
    message: string,
    readonly code: 'unknown-root' | 'absolute-path-rejected' | 'traversal-rejected' | 'symlink-escape',
  ) {
    super(message)
    this.name = 'FilesystemPathError'
  }
}

export class FilesystemCapabilityError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'FilesystemCapabilityError'
  }
}

export interface FilesystemAdapterOptions {
  /** Named allowed root directories (absolute paths). Every call names one root explicitly. */
  readonly roots: Readonly<Record<string, string>>
  readonly allowRead?: boolean
  readonly allowWrite?: boolean
}

async function nearestExistingAncestor(candidate: string): Promise<string> {
  let current = candidate
  for (;;) {
    try {
      await fs.stat(current)
      return current
    } catch {
      const parent = path.dirname(current)
      if (parent === current) return current
      current = parent
    }
  }
}

/**
 * Resolves `relativePath` against `rootAbsolute`, rejecting absolute
 * inputs and textual traversal, then verifies (via `realpath` of the
 * nearest existing ancestor) that the resolved location does not escape
 * the root through a symlink.
 */
export async function resolveWithinRoot(rootAbsolute: string, relativePath: string): Promise<string> {
  if (path.isAbsolute(relativePath)) {
    throw new FilesystemPathError(`path must be relative to the configured root, got absolute path "${relativePath}"`, 'absolute-path-rejected')
  }
  const resolvedRoot = path.resolve(rootAbsolute)
  const resolved = path.resolve(resolvedRoot, relativePath)
  if (resolved !== resolvedRoot && !resolved.startsWith(resolvedRoot + path.sep)) {
    throw new FilesystemPathError(`path "${relativePath}" resolves outside its configured root`, 'traversal-rejected')
  }

  const realRoot = await fs.realpath(resolvedRoot).catch(() => resolvedRoot)
  const ancestor = await nearestExistingAncestor(resolved)
  const realAncestor = await fs.realpath(ancestor).catch(() => ancestor)
  if (realAncestor !== realRoot && !realAncestor.startsWith(realRoot + path.sep)) {
    throw new FilesystemPathError(`path "${relativePath}" escapes its configured root via a symlink`, 'symlink-escape')
  }
  return resolved
}

/** Read/write file adapter constrained to explicitly configured, named root directories. */
export class NodeFilesystemAdapter {
  constructor(private readonly options: FilesystemAdapterOptions) {}

  private rootPath(rootName: string): string {
    const rootPath = this.options.roots[rootName]
    if (rootPath === undefined) {
      throw new FilesystemPathError(`unknown filesystem root "${rootName}"`, 'unknown-root')
    }
    return rootPath
  }

  async readFile(rootName: string, relativePath: string): Promise<string> {
    if (this.options.allowRead === false) {
      throw new FilesystemCapabilityError('read capability is not enabled for this adapter')
    }
    const absolute = await resolveWithinRoot(this.rootPath(rootName), relativePath)
    return fs.readFile(absolute, 'utf8')
  }

  async writeFile(rootName: string, relativePath: string, contents: string): Promise<void> {
    if (!this.options.allowWrite) {
      throw new FilesystemCapabilityError('write capability is not enabled for this adapter')
    }
    const absolute = await resolveWithinRoot(this.rootPath(rootName), relativePath)
    await fs.mkdir(path.dirname(absolute), { recursive: true })
    await fs.writeFile(absolute, contents, 'utf8')
  }

  async exists(rootName: string, relativePath: string): Promise<boolean> {
    const absolute = await resolveWithinRoot(this.rootPath(rootName), relativePath)
    return fs
      .stat(absolute)
      .then(() => true)
      .catch(() => false)
  }

  async remove(rootName: string, relativePath: string): Promise<void> {
    if (!this.options.allowWrite) {
      throw new FilesystemCapabilityError('write capability is not enabled for this adapter')
    }
    const absolute = await resolveWithinRoot(this.rootPath(rootName), relativePath)
    await fs.rm(absolute, { force: true })
  }
}

File: runtimes/python/src/engineering_ui_capabilities_runtime/adapters/filesystem.py
"""Filesystem adapter foundation (§15.3 "Filesystem adapters use configured
roots and explicit read/write capabilities" + "All target paths are
repository-relative, normalized, and checked against traversal and symlink
escape").
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Union


class PathTraversalError(ValueError):
    """A requested relative path resolves outside the configured root
    (either via `..` segments or a symlink)."""


class FilesystemCapabilityError(PermissionError):
    """The adapter was not configured for the requested operation
    (read/write)."""


@dataclass(frozen=True)
class FilesystemCapability:
    can_read: bool = True
    can_write: bool = False


class FilesystemAdapter:
    """All paths passed to this adapter are relative to a single configured
    root; the resolved, real (symlink-following) path must remain under
    that root's resolved real path, or `PathTraversalError` is raised.
    """

    def __init__(self, root: Union[str, Path], capability: FilesystemCapability = FilesystemCapability()) -> None:
        self._root = Path(root).resolve()
        self._capability = capability

    @property
    def root(self) -> Path:
        return self._root

    def _resolve(self, relative_path: str) -> Path:
        if Path(relative_path).is_absolute():
            raise PathTraversalError(f"{relative_path!r} must be relative to the adapter's root, not absolute")
        candidate = (self._root / relative_path).resolve()
        try:
            candidate.relative_to(self._root)
        except ValueError as exc:
            raise PathTraversalError(
                f"{relative_path!r} resolves to {candidate}, which escapes the configured root {self._root}"
            ) from exc
        return candidate

    def exists(self, relative_path: str) -> bool:
        return self._resolve(relative_path).exists()

    def read_text(self, relative_path: str, encoding: str = "utf-8") -> str:
        if not self._capability.can_read:
            raise FilesystemCapabilityError("This adapter instance is not configured for read access")
        return self._resolve(relative_path).read_text(encoding=encoding)

    def read_bytes(self, relative_path: str) -> bytes:
        if not self._capability.can_read:
            raise FilesystemCapabilityError("This adapter instance is not configured for read access")
        return self._resolve(relative_path).read_bytes()

    def write_text(self, relative_path: str, content: str, encoding: str = "utf-8") -> None:
        if not self._capability.can_write:
            raise FilesystemCapabilityError("This adapter instance is not configured for write access")
        target = self._resolve(relative_path)
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(content, encoding=encoding)

    def write_bytes(self, relative_path: str, content: bytes) -> None:
        if not self._capability.can_write:
            raise FilesystemCapabilityError("This adapter instance is not configured for write access")
        target = self._resolve(relative_path)
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_bytes(content)
